src/libsystemd-network/dhcp-option.c

```diff
@@ -93,32 +112,75 @@ static int option_append(uint8_t options[], size_t size, size_t *offset,
+                        r = dhcp_option_append_tlv(options, size, offset, p->option, p->length, p->data);
+                        assert(r >= 0);
+                }
+                break;
+        }
+        case SD_DHCP_OPTION_RELAY_AGENT_INFORMATION: {
+                sd_dhcp_server *server = (sd_dhcp_server *) optval;
+                size_t current_offset = *offset + 2;
+
+                if (server->agent_circuit_id) {
+                        size_t length = strlen(server->agent_circuit_id);
+                        int subcode = SD_DHCP_RELAY_AGENT_CIRCUIT_ID;
+                        r = dhcp_option_append_tlv(options, size, &current_offset, subcode, length, server->agent_circuit_id);
+                        if (r < 0)
+                                return r;
+                }
+                if (server->agent_remote_id) {
+                        size_t length = strlen(server->agent_remote_id);
+                        int subcode = SD_DHCP_RELAY_AGENT_REMOTE_ID;
+                        r = dhcp_option_append_tlv(options, size, &current_offset, subcode, length, server->agent_remote_id);
+                        if (r < 0)
+                                return r;
                 }
 
+                options[*offset] = code;
+                options[*offset + 1] = current_offset - *offset;
+                assert(options[*offset + 1] <= 255);
+                *offset = current_offset;
                 break;
         }
         default:
-                if (*offset + 2 + optlen > size)
-                        return -ENOBUFS;
+                return dhcp_option_append_tlv(options, size, offset, code, optlen, optval);
+        }
+        return 0;
+}
 
-                options[*offset] = code;
-                options[*offset + 1] = optlen;
+int dhcp_option_remove_option(uint8_t *options, size_t length, uint8_t option_code) {
+        assert(options);
 
-                memcpy_safe(&options[*offset + 2], optval, optlen);
-                *offset += 2 + optlen;
+        for (size_t offset = 0; offset < length;) {
+                uint8_t code = options[offset ++];
+                size_t len;
 
-                break;
-        }
+                switch (code) {
+                case SD_DHCP_OPTION_PAD:
+                        continue;
 
-        return 0;
+                case SD_DHCP_OPTION_END:
+                        return -ENOENT;
+                }
+
+                if (length < offset + 1)
+                        return -ENOBUFS;
+
+                len = options[offset ++];
+                if (length < offset + len)
+                        return -ENOBUFS;
+                offset += len;
+
+                if (code == option_code) {
+                        size_t option_offset = offset - len - 2;
+                        memmove(options + option_offset, options + offset, length - offset);
+                        return length - len - 2;
+                }
+        }
+        return -ENOENT;
 }
```

Comment:
this function really could need some unit test

Reply:
How it's better to be done? It's static to the source file now

Reply:
Simply expose the function, and add tests into `test-dhcp-option.c`.

Reply:
Ok(I was wrong about it being static)